ref/haskell/src/Lib.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8)
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Map.HashMap Char Word8
+charsetMap = Map.fromList $ zip charset [0..]
+
+charsetMapInv :: Map.HashMap Word8 Char
+charsetMapInv = Map.fromList $ zip [0..] charset
+
+bech32Polymod :: Data -> Int
+bech32Polymod = foldl' go 1
+  where
+    go chk value = foldl' xor chk' [(- (top .>>. i) .&. 1) .&. g | (g, i) <- zip generator [0..]]
+      where
+        generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3]
+        top = chk .>>. 25
+        chk' = (chk .&. 0x1ffffff) .<<. 5 `xor` (fromIntegral value)
+
+bech32HRPExpand :: HRP -> Data
+bech32HRPExpand hrp = map (.>>. 5) (BS.unpack hrp) ++ [0] ++ map (.&. 31) (BS.unpack hrp)
+
+bech32CreateChecksum :: HRP -> Data -> Data
+bech32CreateChecksum hrp dat = [fromIntegral $ polymod .>>. (5 * (5 - i)) .&. 31 | i <- [0..5]]
+  where
+    values = bech32HRPExpand hrp ++ dat
+    polymod = bech32Polymod (values ++ [0, 0, 0, 0, 0, 0]) `xor` 1
+
+bech32VerifyChecksum :: HRP -> Data -> Bool
+bech32VerifyChecksum hrp dat = bech32Polymod (bech32HRPExpand hrp ++ dat) == 1
+
+bech32Encode :: HRP -> Data -> Maybe BS.ByteString
+bech32Encode hrp dat = do
+    let dat' = dat ++ bech32CreateChecksum hrp dat
+    rest <- mapM (flip Map.lookup charsetMapInv) dat'
+    return $ BSC.concat [hrp, "1", BSC.pack rest]
+
+bech32Decode :: BS.ByteString -> Maybe (HRP, Data)
+bech32Decode bech32 = do
+    guard $ BS.length bech32 <= 90
+    guard $ BSC.map toUpper bech32 == bech32 || BSC.map toLower bech32 == bech32
```

Comment:
It seems that toLower/toUpper will map some non-ASCII characters to ASCII characters (in particular 304, 305, 383, and 8490), so I worry that this may cause those to be unintentionally accepted.